### PR TITLE
Add checks for unit id conflicts.

### LINKED ISSUE #4256: damlc does not detect unit id collisions
In a project where you have:

- package `a`
- package `b` which depends on a but built against an older version of `a`
- package `c` which depends on `a` and `b`

damlc will silently accept this and references to `a` will be written somewhat randomly to one of the DALF packages.

We should detect this and fail loudly.

### PR DESCRIPTION
**Pull request by associahedron**
Fixes #4256 

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs [hunk 1]
```diff
@@ -450,27 +450,35 @@ generateDocTestModuleRule =
...
+                        | "daml-stdlib" `isPrefixOf` baseName = (stringToUnitId baseName, pkg)
+                        | otherwise = ( stringToUnitId $ parseUnitId baseName $ LF.dalfPackageId pkg, pkg)
+                pure (fmap getUnitId dalfPkgOrErr)
+
+    let unitIdConflicts = Map.filter ((>=2) . Set.size) . Map.fromListWith Set.union $
+            [ (unitId, Set.singleton (LF.dalfPackageId dalfPkg))
+            | (unitId, dalfPkg) <- pkgs ]
+    when (not $ Map.null unitIdConflicts) $ do
+        fail $ "Transitive dependencies with same unit id but conflicting package ids: "
+            ++ intercalate ", " (map show $ Map.keys unitIdConflicts)
```
Comment (cocreature) [inline compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs]: Should we include the the package ids in the error? Seems potentially useful for debugging.

File: compiler/damlc/tests/src/DA/Test/Packaging.hs [hunk 2]
```diff
@@ -431,6 +431,152 @@ tests damlc repl davlDar = testGroup "Packaging" $
...
+          -- Check that two pacages with the same unit id is flagged as an error.
+          let projA = projDir </> "a"
+          let projB = projDir </> "b"
+          let projC = projDir </> "c"
+
+          createDirectoryIfMissing True (projA </> "src")
+          createDirectoryIfMissing True (projB </> "src")
+          createDirectoryIfMissing True (projC </> "src")
+
+          writeFileUTF8 (projA </> "daml.yaml") $ unlines
```
Comment (cocreature) [inline compiler/damlc/tests/src/DA/Test/Packaging.hs]: We should really make a helper function for creating `daml.yaml` files in tests :slightly_smiling_face:

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs [hunk 3]
```diff
@@ -155,6 +155,16 @@ createProjectPackageDb opts thisSdkVer deps dataDeps = do
...
         pure (pkgId, package, dalf, unitId)
 
+    let unitIdConflicts = MS.filter ((>=2) . Set.size) .  MS.fromListWith Set.union $ concat
+            [ [ (unitId, Set.singleton pkgId)
+              | (pkgId, _package, _dalf, unitId) <- pkgs ]
+            , [ (unitId, Set.singleton (LF.dalfPackageId dalfPkg))
+              | (unitId, dalfPkg) <- MS.toList dependencies ]
+            ]
+    when (not $ MS.null unitIdConflicts) $ do
+        fail $ "Transitive dependencies with same unit id but conflicting package ids: "
```
Comment (cocreature) [inline compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs]: Same here package ids might be useful.

File: compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs (hunk 1 above)
Comment (associahedron) [reply compiler/damlc/daml-ide-core/src/Development/IDE/Core/Rules/Daml.hs]: Sounds good, added

File: compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs (hunk 3 above)
Comment (associahedron) [reply compiler/damlc/lib/DA/Cli/Damlc/Packaging.hs]: Added here too

**Pull request merged**